### PR TITLE
Webservices tests : CRUD Stores

### PR DESCRIPTION
**Pull request by davidPrestashop**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Webservices tests : CRUD Stores
| Type?             | refacto
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI 🟢 + automated tests

### CODE REVIEW

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 1]
```diff
@@ -125,6 +128,111 @@ class AddStore extends BOBasePage {
     // Return successful message
     return this.getAlertSuccessBlockParagraphContent(page);
   }
+
+  async getInputValue(page: Page, inputName: string, languageId?: string): Promise<string> {
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:138]: Please add JS doc

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 2]
```diff
@@ -125,6 +128,111 @@ class AddStore extends BOBasePage {
...
+      default:
+        throw new Error(`Column ${inputName} was not found`);
+    }
+
+    const result = await this.getAttributeContent(page, selector, 'value');
+
+    return result;
+  }
+
+  async getSelectValue(page: Page, name: string): Promise<string> {
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:221]: Same here

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 3]
```diff
@@ -125,6 +128,111 @@ class AddStore extends BOBasePage {
...
+      default:
+        throw new Error(`Field ${name} was not found`);
+    }
+
+    const result = page.$eval(selector, (node: HTMLSelectElement) => node.value);
+
+    return result;
+  }
+
+  async isActive(page: Page, toggle: string): Promise<boolean> {
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:243]: Same here

File: tests/UI/pages/BO/shopParameters/stores/add.ts (hunk 1 above)
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:138]: Can you add JSDoc ?

File: tests/UI/pages/BO/shopParameters/stores/add.ts (hunk 2 above)
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:221]: Can you add JSDoc ?

File: tests/UI/pages/BO/shopParameters/stores/add.ts (hunk 3 above)
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts:243]: Can you add JSDoc ?

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 4]
```diff
@@ -125,6 +128,111 @@ class AddStore extends BOBasePage {
...
+    const result = this.isChecked(page, this.statusToggle(toggle));
+
+    return result;
```
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
    return this.isChecked(page, this.statusToggle(toggle));
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 5]
```diff
@@ -0,0 +1,1114 @@
+// Import utils
+import helper from '@utils/helpers';
+import testContext from '@utils/testContext';
+import storeXml from '@webservices/store/storeXml';
+
+// Import webservices
+import StoreWS from '@webservices/store/storeWS';
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:7]:
```suggestion
import storeXml from '@webservices/store/storeXml';
import StoreWS from '@webservices/store/storeWS';
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 6]
```diff
@@ -0,0 +1,1114 @@
+// Import utils
+import helper from '@utils/helpers';
+import testContext from '@utils/testContext';
+import storeXml from '@webservices/store/storeXml';
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 7]
```diff
@@ -0,0 +1,1114 @@
...
+      let apiResponse : APIResponse;
+
+      it('should check response status', async function () {
+        await testContext.addContextItem(this, 'testIdentifier', 'requestGetBlankStatus', baseContext);
+
+        apiResponse = await StoreWS.getBlank(
+          apiContext,
+          authorization,
+        );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 8]
```diff
@@ -0,0 +1,1114 @@
...
+      let apiResponse : APIResponse;
+
+      it('should check response status', async function () {
+        await testContext.addContextItem(this, 'testIdentifier', 'requestGetSynopsisStatus', baseContext);
+
+        apiResponse = await StoreWS.getSynopsis(
+          apiContext,
+          authorization,
+        );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 9]
```diff
@@ -0,0 +1,1114 @@
...
+      let storesNode: Element[];
+
+      it('should check response status', async function () {
+        await testContext.addContextItem(this, 'testIdentifier', 'requestGetStatus1', baseContext);
+
+        apiResponse = await StoreWS.getAll(
+          apiContext,
+          authorization,
+        );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:244]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 10]
```diff
@@ -0,0 +1,1114 @@
...
+        it('should check response status', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'requestPostStatus1', baseContext);
+
+          apiResponse = await StoreWS.add(
+            apiContext,
+            authorization,
+            xmlCreate,
+          );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:306]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 11]
```diff
@@ -0,0 +1,1114 @@
...
+            if (oNode.nodeName === 'id') {
+              expect(oNode.textContent).to.be.eq(storeNodeID);
+            } else if (oNode.nodeName === 'name' || oNode.nodeName === 'hours' || oNode.nodeName === 'address1'
+              || oNode.nodeName === 'address2' || oNode.nodeName === 'note') {
+              const objectNodeValueEN = storeXml.getLangEltTextContent(
+                xmlResponse,
+                oNode.nodeName,
+                '1',
+              );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:388]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 12]
```diff
@@ -0,0 +1,1114 @@
...
+              const createNodeValueEN = storeXml.getLangEltTextContent(
+                xmlCreate,
+                oNode.nodeName,
+                '1',
+              );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:394]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 13]
```diff
@@ -0,0 +1,1114 @@
...
+          const valueIDCountry = await addStorePage.getSelectValue(page, 'id_country');
+
+          expect(valueIDCountry).to.be.eq(xmlValueIDCountry);
+        });
+
+        it('should check store\'s state', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreState', baseContext);
+
+          // const xmlValueIDState = storeXml.getEltTextContent(xmlCreate, 'id_state');
+          const xmlValueIDState = '';
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]: Why the first line is commented ?

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 14]
```diff
@@ -0,0 +1,1114 @@
...
+
+          const xmlValueFax = storeXml.getEltTextContent(xmlCreate, 'email');
+          const valueFax = await addStorePage.getInputValue(page, 'email');
+          expect(valueFax).to.be.eq(xmlValueFax);
+        });
+
+        it('should check store is active', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreActive', baseContext);
+
+          const xmlValueActive: string = storeXml.getEltTextContent(xmlCreate, 'active');
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
          const xmlValueActive = storeXml.getEltTextContent(xmlCreate, 'active');
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 15]
```diff
@@ -0,0 +1,1114 @@
...
+          expect(dayValue).to.be.eq(dayHours.substring(2, dayHours.length - 3));
+        });
+
+        it('should check store\'s sunday hours language 2', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreHoursSundayLang2', baseContext);
+
+          const dayHours = hoursArrLang2[6];
+          const dayValue = await addStorePage.getInputValue(page, 'sunday', '2');
+          expect(dayValue).to.be.eq(dayHours.substring(2, dayHours.length - 3));
+        });
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]: Could you refacto this code and loop on a weeksday array ?

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 16]
```diff
@@ -0,0 +1,1114 @@
...
+        it(`should check response status of ${StoreWS.endpoint}/{id}`, async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'requestPutStatus1', baseContext);
+          xmlUpdate = getUpdateStoreXml(storeNodeID);
+          apiResponse = await StoreWS.update(
+            apiContext,
+            authorization,
+            storeNodeID,
+            xmlUpdate,
+          );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:666]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 17]
```diff
@@ -0,0 +1,1114 @@
...
+        it('should check response status', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'requestGetPutStatus2', baseContext);
+
+          apiResponse = await StoreWS.getById(
+            apiContext,
+            authorization,
+            storeNodeID,
+          );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:707]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 18]
```diff
@@ -0,0 +1,1114 @@
...
+            if (oNode.nodeName === 'id') {
+              expect(oNode.textContent).to.be.eq(storeNodeID);
+            } else if (oNode.nodeName === 'name' || oNode.nodeName === 'hours' || oNode.nodeName === 'address1'
+              || oNode.nodeName === 'address2' || oNode.nodeName === 'note') {
+              const objectNodeValueEN = storeXml.getLangEltTextContent(
+                xmlResponse,
+                oNode.nodeName,
+                '1',
+              );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:749]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 19]
```diff
@@ -0,0 +1,1114 @@
...
+              const createNodeValueEN = storeXml.getLangEltTextContent(
+                xmlUpdate,
+                oNode.nodeName,
+                '1',
+              );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:755]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 20]
```diff
@@ -0,0 +1,1114 @@
...
+        });
+
+        it('should check store\'s sunday hours language 2', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreHoursSundayLang22', baseContext);
+          const xmlValueUpdatedHoursLang2 = storeXml.getLangEltTextContent(xmlUpdate, 'hours', '2');
+          const hoursArrUpdatedLang2: string[] = xmlValueUpdatedHoursLang2.split(',');
+          const dayHours = hoursArrUpdatedLang2[6];
+          const dayValue = await addStorePage.getInputValue(page, 'sunday', '2');
+          expect(dayValue).to.be.eq(dayHours.substring(2, dayHours.length - 3));
+        });
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]: Please create a loop

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 21]
```diff
@@ -0,0 +1,1114 @@
...
+    describe(`Endpoint : ${StoreWS.endpoint} - Method : DELETE `, () => {
+      it(`should request the endpoint ${StoreWS.endpoint}/{id} with method DELETE`, async function () {
+        await testContext.addContextItem(this, 'testIdentifier', 'requestEndpointStoresMethodDelete', baseContext);
+
+        const apiResponse = await StoreWS.delete(
+          apiContext,
+          authorization,
+          storeNodeID,
+        );
+
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts:1004]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks.

Thanks for the contribution @davidPrestashop

**Review by boherm — DISMISSED** [review]

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 22]
```diff
@@ -125,6 +128,121 @@ class AddStore extends BOBasePage {
...
+        throw new Error(`Column ${inputName} was not found`);
+    }
+
+    return await this.getAttributeContent(page, selector, 'value');
+  }
+
+  /**
+   *
+   * @param page
+   * @param name
```
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
   * @param page {Page}
   * @param name {string}
```

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 23]
```diff
@@ -125,6 +128,121 @@ class AddStore extends BOBasePage {
     // Return successful message
     return this.getAlertSuccessBlockParagraphContent(page);
   }
+
+  /**
+   *
+   * @param page
+   * @param inputName
+   * @param languageId
```
Comment (Progi1984) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
   * @param page {Page}
   * @param inputName {string}
   * @param languageId {string|undefined}
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 24]
```diff
@@ -0,0 +1,1047 @@
...
+      methods: ['all'],
+    },
+  ];
+  const xmlCreate: string = getStoreXml();
+  const xmlValueHoursLang1 = storeXml.getLangEltTextContent(xmlCreate, 'hours', '1');
+  const hoursArrLang1: string[] = xmlValueHoursLang1.split(',');
+  const xmlValueHoursLang2 = storeXml.getLangEltTextContent(xmlCreate, 'hours', '2');
+  const hoursArrLang2: string[] = xmlValueHoursLang2.split(',');
+  let xmlUpdate: string;
+  // const week = [{day: 'monday'}, {day: 'tuesday'}, {day: 'wednesday'}, {day: 'thursday'}, {day: 'friday'}, {day: 'saturday'}, {day: 'sunday'}];
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 25]
```diff
@@ -0,0 +1,1047 @@
...
+
+          const xmlValueIDCountry = storeXml.getEltTextContent(xmlUpdate, 'id_country');
+          const valueIDCountry = await addStorePage.getSelectValue(page, 'id_country');
+          expect(valueIDCountry).to.be.eq(xmlValueIDCountry);
+        });
+
+        it('should check store\'s state', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreState2', baseContext);
+
+          // const xmlValueIDState = storeXml.getEltTextContent(xmlUpdate, 'id_state');
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks

**Comment by Progi1984** [comment]
UI tests : https://github.com/Progi1984/ga.tests.ui.pr/actions/runs/5646448448

**Review by boubkerbribri — DISMISSED** [review]

File: tests/UI/campaigns/functional/WS/01_storesCRUD.ts [hunk 26]
```diff
@@ -0,0 +1,1046 @@
...
+          const xmlValueIDCountry = storeXml.getEltTextContent(xmlCreate, 'id_country');
+          const valueIDCountry = await addStorePage.getSelectValue(page, 'id_country');
+
+          expect(valueIDCountry).to.be.eq(xmlValueIDCountry);
+        });
+
+        it('should check store\'s state', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkStoreState', baseContext);
+
+          const xmlValueIDState  = storeXml.getEltTextContent(xmlCreate, 'id_state');
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/WS/01_storesCRUD.ts]:
```suggestion
          const xmlValueIDState = storeXml.getEltTextContent(xmlCreate, 'id_state');
```

**Review by boubkerbribri — DISMISSED** [review]

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 27]
```diff
@@ -125,6 +128,121 @@ class AddStore extends BOBasePage {
     // Return successful message
     return this.getAlertSuccessBlockParagraphContent(page);
   }
+
+  /**
+   *
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
   * Get input value
```

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 28]
```diff
@@ -125,6 +128,121 @@ class AddStore extends BOBasePage {
...
+        break;
+      default:
+        throw new Error(`Column ${inputName} was not found`);
+    }
+
+    return this.getAttributeContent(page, selector, 'value');
+  }
+
+  /**
+   *
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
   * Get select value
```

File: tests/UI/pages/BO/shopParameters/stores/add.ts [hunk 29]
```diff
@@ -125,6 +128,121 @@ class AddStore extends BOBasePage {
...
+        break;
+      default:
+        throw new Error(`Field ${name} was not found`);
+    }
+
+    return page.$eval(selector, (node: HTMLSelectElement) => node.value);
+  }
+
+  /**
+   *
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/shopParameters/stores/add.ts]:
```suggestion
   * Get status
```

**Review by nesrineabdmouleh — COMMENTED** [review]
Small feedback

**Review by nesrineabdmouleh — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by boubkerbribri** [comment]
Thanks @davidPrestashop

**Pull request merged**